File: python/problems/leetcode/0104_maximum_depth_of_binary_tree/solution.py
"""LeetCode 104: Maximum Depth of Binary Tree."""

from __future__ import annotations

from typing import Optional


class TreeNode:
    def __init__(
        self,
        val: int = 0,
        left: Optional["TreeNode"] = None,
        right: Optional["TreeNode"] = None,
    ) -> None:
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        return 1 + max(self.maxDepth(root.left), self.maxDepth(root.right))


if __name__ == "__main__":
    root = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
    assert Solution().maxDepth(root) == 3
    assert Solution().maxDepth(None) == 0
    print("0104 OK")

File: cpp/problems/leetcode/0104_maximum_depth_of_binary_tree/solution.cpp
// LeetCode 104
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
    TreeNode(int v, TreeNode* l, TreeNode* r) : val(v), left(l), right(r) {}
};

class Solution {
public:
    int maxDepth(TreeNode* root) {
        if (!root) return 0;
        return 1 + max(maxDepth(root->left), maxDepth(root->right));
    }
};

int main() {
    TreeNode n15(15), n7(7), n9(9), n20(20, &n15, &n7);
    TreeNode root(3, &n9, &n20);
    assert(Solution().maxDepth(&root) == 3);
    assert(Solution().maxDepth(nullptr) == 0);
    cout << "0104 OK" << endl;
    return 0;
}
